pydra/engine/task.py

```diff
@@ -0,0 +1,443 @@
+    """
+    return NotImplementedError
+
+
+def now():
+    return dt.datetime.utcnow().isoformat(timespec='microseconds')
+
+
+prov_context = "https://openprovenance.org/prov.jsonld"
+schema_context = "https://schema.org/docs/jsonldcontext.json"
+pydra_context = {"pydra": "https://uuid.pydra.org/"}
+
+# audit flags
+AUDIT_PROV = 0x01
+AUDIT_RESOURCE = 0x02
+
+
+def gen_uuid():
+    import uuid
+    return uuid.uuid4().hex
+
+
+class Messenger:
+    @abc.abstractmethod
+    def send(self, message, **kwargs):
+        pass
+
+
+class PrintMessenger(Messenger):
+    
+    def send(self, message, **kwargs):
+        import json
+        mid = gen_uuid()
+        print('id: {0}\n{1}'.format(mid,
+                                    json.dumps(message, ensure_ascii=False, 
+                                               indent=2, sort_keys=False)))
+
+
+class FileMessenger(Messenger):
+    
+    def send(self, message, **kwargs):
+        import json
+        mid = gen_uuid()
+        with open(os.path.join(kwargs['message_dir'], 
+                               mid + '.jsonld'), 'wt') as fp:
+            json.dump(message, fp, ensure_ascii=False, indent=2, 
+                      sort_keys=False)
+
+
+class RemoteRESTMessenger(Messenger):
+
+    def send(self, message, **kwargs):
+        import requests
+        r = requests.post(kwargs['post_url'], json=message, 
+                          auth=kwargs['auth']() if getattr(kwargs['auth'], 
+                                                           '__call__', None) 
+                                                else kwargs['auth'])
+        return r.status_code
+
+
+def send_message(message, messengers=None, **kwargs):
+    """Send nidm messages for logging provenance and auditing
+    """
+    for messenger in messengers:
+        messenger.send(message, **kwargs)
+
+
+def make_message(obj, context="https://schema.pydra.org/context.jsonld"):
+    message = {"@context": context}
+    message.update(**obj)
+    return message
+
+
+@dc.dataclass
+class RuntimeSpec:
+    outdir: ty.Optional[str] = None
+    container: ty.Optional[str] = 'shell'
+    network: bool = False
+    """
+    from CWL:
+    InlineJavascriptRequirement
+    SchemaDefRequirement
+    DockerRequirement
+    SoftwareRequirement
+    InitialWorkDirRequirement
+    EnvVarRequirement
+    ShellCommandRequirement
+    ResourceRequirement
+    
+    InlineScriptRequirement
+    """
+
+
+@dc.dataclass
+class BaseSpec:
+    @property
+    def hash(self):
+        return sha256(str(self).encode()).hexdigest()
+
+
+@dc.dataclass
+class Result:
+    output: ty.Optional[ty.Any] = None
+
+
+class BaseTask:
+    """This is a base class for Task objects.
+    """
+
+    _api_version: str = "0.0.1"  # Should generally not be touched by subclasses
+    _task_version: ty.Optional[str] = None  # Task writers encouraged to define and increment when implementation changes sufficiently
+    _version: str  # Version of tool being wrapped
```

Comment:
Should this be optional, or potentially a range of versions? Or is the idea that tasks *should* be highly constrained to a specific version, and my way of indicating that this task still works for a new version is:

```Python
class FLIRT6(FLIRT509):
    _version = '6.0.0
```